crates/rpcsrv/src/flight_handler.rs

```diff
@@ -0,0 +1,324 @@
+use crate::errors::{Result, RpcsrvError};
+
+use dashmap::DashMap;
+use datafusion::arrow::ipc::writer::IpcWriteOptions;
+use datafusion_ext::vars::SessionVars;
+use datafusion_proto::protobuf::PhysicalPlanNode;
+use once_cell::sync::Lazy;
+use sqlexec::{
+    context::remote::RemoteSessionContext,
+    engine::{Engine, SessionStorageConfig, TrackedSession},
+    extension_codec::GlareDBExtensionCodec,
+    OperationInfo,
+};
+use std::{pin::Pin, sync::Arc};
+use uuid::Uuid;
+
+pub use arrow_flight::flight_service_server::FlightServiceServer;
+use arrow_flight::{
+    encode::FlightDataEncoderBuilder, flight_service_server::FlightService, sql::*, Action,
+    FlightDescriptor, FlightEndpoint, FlightInfo, IpcMessage, SchemaAsIpc, Ticket,
+};
+use arrow_flight::{
+    sql::{
+        metadata::{SqlInfoData, SqlInfoDataBuilder},
+        server::FlightSqlService,
+    },
+    HandshakeRequest, HandshakeResponse,
+};
+use datafusion_proto::physical_plan::AsExecutionPlan;
+use futures::TryStreamExt;
+use futures::{lock::Mutex, Stream};
+use prost::Message;
+use tonic::{Request, Response, Status, Streaming};
+
+static INSTANCE_SQL_DATA: Lazy<SqlInfoData> = Lazy::new(|| {
+    let mut builder = SqlInfoDataBuilder::new();
+    // Server information
+    builder.append(SqlInfo::FlightSqlServerName, "GlareDB Flight Server");
+    builder.append(SqlInfo::FlightSqlServerVersion, env!("CARGO_PKG_VERSION"));
+    builder.append(SqlInfo::FlightSqlServerArrowVersion, "1.3");
+    builder.build().unwrap()
+});
+
+macro_rules! status {
+    ($desc:expr, $err:expr) => {
+        Status::internal(format!("{}: {} at {}:{}", $desc, $err, file!(), line!()))
+    };
+}
+
+pub struct FlightSessionHandler {
+    engine: Arc<Engine>,
+    /// The remote context is used to execute the queries in a stateless manner.
+    remote_ctx: Arc<RemoteSessionContext>,
+    sessions: DashMap<String, Arc<Mutex<TrackedSession>>>,
+}
+
+impl FlightSessionHandler {
+    async fn create_ctx(&self, handle: &str) -> Result<Arc<Mutex<TrackedSession>>, Status> {
+        if self.sessions.contains_key(handle) {
+            let sess = self.sessions.get(handle).unwrap().clone();
+            return Ok(sess);
+        }
+        let uuid = Uuid::parse_str(handle)
+            .map_err(|e| RpcsrvError::ParseError(format!("Error parsing uuid: {e}")))?;
+        let _ = self
+            .engine
+            .new_remote_session_context(uuid, SessionStorageConfig::default())
+            .await
+            .map_err(RpcsrvError::from)?;
+
+        let sess = self
+            .engine
+            .new_local_session_context(SessionVars::default(), SessionStorageConfig::default())
+            .await
+            .map_err(RpcsrvError::from)?;
+
+        let sess = Arc::new(Mutex::new(sess));
```

Comment:
I would like to understand better what the mutex is protecting: what's the situation when this might see concurrent access, and I can't quite see it from reading the code.

Reply:
`TrackedSession` doesn't utilize any inner mutability, so most methods on it do require `&mut self`. So we gotta wrap it in `Arc<Mutex<>>`. We do the same thing in the py & js bindings. 

`pub(super) type PyTrackedSession = Arc<Mutex<TrackedSession>>;`
`pub(super) type JsTrackedSession = Arc<Mutex<TrackedSession>>;`